### DISCUSSION TITLE
Understand adapters

### DISCUSSION THREAD

**Discussion by mariomnts** [prisma/prisma#22721] [Q&A] [answered]
### Question

Hi:

As part of the discussion in this issue (https://github.com/prisma/prisma/issues/1964) I wanted to better understand how Prisma adapters work and how new ones can be implemented.

Something I've seen all along their code is that they call a client which is not initialised in the adapters (`this.client`).

**Comment by janpio** [comment, marked as answer]
The database client is constructed before you instantiate the adapter and Prisma Client, and then given to the adapter. For Neon for example, it is a `Pool` object:
```
const pool = new Pool({ connectionString });
const adapter = new PrismaNeon(pool);
const prisma = new PrismaClient({ adapter });
```
(from https://www.prisma.io/blog/serverless-database-drivers-KML1ehXORxZV)

And in the code:
https://github.com/prisma/prisma/blob/693260761eecb61faed7f2229e63676be7b2453f/packages/adapter-neon/src/neon.ts#L129-L141
